rust-symcrypt/src/rsa/pss.rs

```diff
@@ -0,0 +1,281 @@
+//! PSS functions for [`RsaKeyPair`] and [`RsaPublicKey`]. For more info please refer to symcrypt.h
+//!
+//! Signing functionality is locked to only be usable by [`RsaKeyPair`].
+//! Verification functionality is provided by both [`RsaKeyPair`] and [`RsaPublicKey`].
+//!
+//! # Example
+//!
+//! ## Sign and Verify with [`RsaKeyPair`]
+//!
+//! ```rust
+//! use symcrypt::rsa::{RsaKeyPair, RsaKeyUsage};
+//! use symcrypt::hash::{sha256, HashAlgorithm, SHA256_RESULT_SIZE};
+//!
+//! // Generate key pair
+//! let key_pair = RsaKeyPair::generate_new(2048, None, RsaKeyUsage::SignAndEncrypt).unwrap();
+//!
+//! // Set up message
+//! let hashed_message = sha256(b"hello world");
+//! let hash_algorithm = HashAlgorithm::Sha256;
+//! let salt_length = SHA256_RESULT_SIZE; // 32 bytes for SHA256
+//!
+//! // Create and verify the signature
+//! let signature = key_pair.pss_sign(&hashed_message, hash_algorithm, salt_length).unwrap();
+//! let verify_result = key_pair.pss_verify(&hashed_message, &signature, hash_algorithm, salt_length);
+//!
+//! assert!(verify_result.is_ok());
+//! ```
+//!
+use crate::errors::SymCryptError;
+use crate::hash::HashAlgorithm;
+use crate::rsa::{RsaKeyPair, RsaPublicKey};
+use crate::NumberFormat;
+use symcrypt_sys::PSYMCRYPT_RSAKEY;
+
+impl RsaKeyPair {
+    /// `pss_sign` is only available for [`RsaKeyPair`].
+    ///
+    /// This function returns a `Vec<u8>` that represents the signature of the hashed message, or a [`SymCryptError`] if the operation failed.
+    ///
+    /// `hashed_message` is a `&[u8]` that represents the message that has been hashed using the hash algorithm specified in `hash_algorithm`.
```

Comment:
NP: my understanding is that for PSS, you can pass in an arbitrarily long message and SymCrypt will hash it for you using the provided hash algorithm. In other words, you don't need to pre-hash the message. Although, confusingly, SymCrypt calls the corresponding parameter `pbHashValue`, and BCrypt will simply truncate inputs that are longer than the modulus (which is a very footgun design IMO... and one that I've blasted myself with before). cc @NielsFerguson to confirm that I'm correct here.

Reply:
For PSS you still pass in the hash results the same as for pkcs.

Take a look at MSCryptRsaSignHash() which calls both.